### ISSUE TITLE
Heuristics for computing max_local_memory are off

### ISSUE THREAD

**Issue by Munksgaard** [diku-dk/futhark#1191]
I have the following program (from LocVolCalib.fut):

```
let tridagPar [n] (a:  [n]f64, b: [n]f64, c: [n]f64, y: [n]f64 ): *[n]f64 =
  #[unsafe]
  ----------------------------------------------------
  -- Recurrence 1: b[i] = b[i] - a[i]*c[i-1]/b[i-1] --
  --   solved by scan with 2x2 matrix mult operator --
  ----------------------------------------------------
  let b0   = b[0]
  let mats = map  (\i ->
                     if 0 < i
                     then (b[i], 0.0-a[i]*c[i-1], 1.0, 0.0)
                     else (1.0,  0.0,             0.0, 1.0))
                  (iota n)
  let scmt = scan (\(a0,a1,a2,a3) (b0,b1,b2,b3) ->
                     let value = 1.0/(a0*b0)
                     in ( (b0*a0 + b1*a2)*value,
                          (b0*a1 + b1*a3)*value,
                          (b2*a0 + b3*a2)*value,
                          (b2*a1 + b3*a3)*value))
                  (1.0,  0.0, 0.0, 1.0) mats
  let b    = map (\(t0,t1,t2,t3) ->
                    (t0*b0 + t1) / (t2*b0 + t3))
                 scmt
  ------------------------------------------------------
  -- Recurrence 2: y[i] = y[i] - (a[i]/b[i-1])*y[i-1] --
  --   solved by scan with linear func comp operator  --
  ------------------------------------------------------
  let y0   = y[0]
  let lfuns= map  (\i  ->
                     if 0 < i
                     then (y[i], 0.0-a[i]/b[i-1])
                     else (0.0,  1.0))
                  (iota n)
  let cfuns= scan (\(a: (f64,f64)) (b: (f64,f64)): (f64,f64)  ->
                     let (a0,a1) = a
                     let (b0,b1) = b
                     in ( b0 + b1*a0, a1*b1 ))
                  (0.0, 1.0) lfuns
  let y    = map (\(tup: (f64,f64)): f64  ->
                    let (a,b) = tup
                    in a + b*y0)
                 cfuns
  ------------------------------------------------------
  -- Recurrence 3: backward recurrence solved via     --
  --             scan with linear func comp operator  --
  ------------------------------------------------------
  let yn   = y[n-1]/b[n-1]
  let lfuns= map (\k  ->
                    let i = n-k-1
                    in  if   0 < k
                        then (y[i]/b[i], 0.0-c[i]/b[i])
                        else (0.0,       1.0))
                 (iota n)
  let cfuns= scan (\(a: (f64,f64)) (b: (f64,f64)): (f64,f64)  ->
                     let (a0,a1) = a
                     let (b0,b1) = b
                     in (b0 + b1*a0, a1*b1))
                  (0.0, 1.0) lfuns
  let y    = map (\(tup: (f64,f64)): f64  ->
                    let (a,b) = tup
                    in a + b*yn)
                 cfuns
  let y    = map (\i -> y[n-i-1]) (iota n)
  in y

let main [m] [n] (as:  [m][n]f64) (bs: [m][n]f64) (cs: [m][n]f64) (ys: [m][n]f64): *[m][n]f64 =
  map4 (\a b c y -> tridagPar (a, b, c, y)) as bs cs ys
```

When I run it on an RTX2080Ti using the reuse-allocations branch of Futhark and a dataset of size 1x769 I get the following error message:

```
$ futhark dataset -b -g [1][769]f64 -g [1][769]f64 -g [1][769]f64 -g [1][769]f64 | ./tridag64-reuse-allocations -D > /dev/null
Using platform: NVIDIA CUDA
Using device: GeForce RTX 2080 Ti
Lockstep width: 32
Default group size: 256
Default number of groups: 272
Creating OpenCL program...
OpenCL compiler options: -DLOCKSTEP_WIDTH=32 -Dmainzisegmap_group_sizze_9431=256 -Dmainzisegmap_group_sizze_9490=256 -Dmainzisegmap_group_sizze_9570=256 -Dmainzisegmap_group_sizze_9624=256 -Dmainzisegmap_group_sizze_9778=256 -Dmainzisegmap_num_groups_9572=272 -Dmainzisegscan_group_sizze_9537=256 -Dmainzisegscan_group_sizze_9671=256 -Dmainzisegscan_group_sizze_9883=256 -Dmainzisegscan_num_groups_9539=272 -Dmainzisegscan_num_groups_9673=272 -Dmainzisegscan_num_groups_9885=272 -Dmainzisuff_intra_par_1=32
Building OpenCL program...
Created kernel main.scan_stage1_9543.
Created kernel main.scan_stage1_9677.
Created kernel main.scan_stage1_9889.
Created kernel main.scan_stage2_9543.
Created kernel main.scan_stage2_9677.
Created kernel main.scan_stage2_9889.
Created kernel main.scan_stage3_9543.
Created kernel main.scan_stage3_9677.
Created kernel main.scan_stage3_9889.
Created kernel main.segmap_9428.
Created kernel main.segmap_9487.
Created kernel main.segmap_9568.
Created kernel main.segmap_9621.
Created kernel main.segmap_9775.
Created kernel main.segmap_intragroup_8752.
Allocating 6152 bytes for arr->mem in space 'device' (then allocated: 6152 bytes) (new peak).
Actually allocating the desired block.
Allocating 6152 bytes for arr->mem in space 'device' (then allocated: 12304 bytes) (new peak).
Actually allocating the desired block.
Allocating 6152 bytes for arr->mem in space 'device' (then allocated: 18456 bytes) (new peak).
Actually allocating the desired block.
Allocating 6152 bytes for arr->mem in space 'device' (then allocated: 24608 bytes) (new peak).
Actually allocating the desired block.
n: 769, max_group_size: 1024
Compared 32 <= 769.
bytes: 30760, local_memory_capacity: 49136, intra_suff_and_fits: true
Allocating 6152 bytes for mem_10354 in space 'device' (then allocated: 30760 bytes) (new peak).
Actually allocating the desired block.
Launching main.segmap_intragroup_8752 with global work size [769] and local work size [769]; local memory parameters sum to 30760 bytes.
./tridag64-reuse-allocations: tridag64-reuse-allocations.c:5762: OpenCL call
  clEnqueueNDRangeKernel(ctx->opencl.queue, ctx->mainzisegmap_intragroup_8752, 1, ((void *)0), global_work_sizze_10807, local_work_sizze_10811, 0, ((void *)0), ctx->profiling_paused || !ctx->profiling ? ((void *)0) : opencl_get_event(&ctx->opencl, &ctx->mainzisegmap_intragroup_8752_runs, &ctx->mainzisegmap_intragroup_8752_total_runtime))
failed with error code -5 (Out of resources)
```

(I've inserted the line showing local_memory_capacity.)

So, we are only using 30760 bytes out of an estimated 49136 available, yet we still get an out of resources error?

The rts code for opencl notes the following:

``` c
  // The OpenCL implementation may reserve some local memory bytes for
  // various purposes.  In principle, we should use
  // clGetKernelWorkGroupInfo() to figure out for each kernel how much
  // is actually available, but our current code generator design
  // makes this infeasible.  Instead, we have this nasty hack where we
  // arbitrarily subtract some bytes, based on empirical measurements
  // (but which might be arbitrarily wrong).  Fortunately, we rarely
  // try to really push the local memory usage.
```

But I suspect something else may be wrong?

**Comment by Munksgaard** [comment]
To be clear, it works fine when using an input of size 1x768 instead.

My question is, why is there such a big discrepancy between what we think is the
available memory and what is actually the available memory?

**Comment by athas** [comment]
I can't imagine the problem referenced by the comment is the issue here.  I guess you could modify the generated code to call `clGetKernelWorkGroupInfo()` and see whether it is!  But this must be something else.  Maybe we compute the memory requirements incorrectly?

**Comment by Munksgaard** [comment]
> Maybe we compute the memory requirements incorrectly?

When running the example program with an input size of 1x768, the reported peak
memory usage matches the computed memory usage.

> I guess you could modify the generated code to call
> `clGetKernelWorkGroupInfo()` and see whether it is!

I've been trying to do this, but it doesn't seem to work... Here's what I'm
trying to do:

``` opencl
    cl_ulong local_mem_size;
    OPENCL_SUCCEED_OR_RETURN(clGetKernelWorkGroupInfo(ctx->mainzisegmap_intragroup_8752,
                                                      ctx->opencl.device, CL_KERNEL_LOCAL_MEM_SIZE,
                                                      sizeof(cl_ulong), &local_mem_size, NULL));

```

But the result is just 1. Am I holding it wrong?

**Comment by Munksgaard** [comment]
Aha, by moving the call to `clGetKernelWorkGroupInfo` to a later point, I get
something that looks more realistic: 30800 bytes. It's a bit more than the 30760
we calculate, but it's still far from the 49136 byte capacity we expect.

**Comment by athas** [comment]
To me that implies that the local memory usage is not the problem.  Could you modify the code to shrink the amount of local memory it asks for?  It'll be in one of the `clSetKernelArg()` calls.  This will make the kernel malfunction of course, but it should launch, if the problem is that we ask for too much local memory.

**Comment by athas** [comment]
It works for me on `master`.  Given the name of the executable, my guess is that you are debugging your memory merging pass.  On `master`, the intra-group version is not picked, *precisely* because it uses too much memory.  My hypothesis is that your pass is partly functional, in that it fools the compiler into thinking that it doesn't need all that much memory, but for some reason we still allocate the old (and too large) amount.  I have no idea how that might happen.

**Comment by athas** [comment]
I suspect it's actually running out of registers here.

**Comment by athas** [comment]
Yes, if I add `__attribute__((reqd_work_group_size(769, 1, 1)))` to the kernel, then it will run.  Bummer.

**Comment by athas** [comment]
Fixed in 9895f9167978a2acedafebcb39cd32f897a59798 (the local memory part).

**Comment by athas** [comment]
I have encountered the same issue with a different program (tpacf) on OpenCL, while CUDA works fine.
